File: LLD/Problems/ATM/Java/ATMService.java
package Problems.ATM;

public class ATMService {
    private final Bank bank;
    private final ATM atm;

    public ATMService(Bank bank, ATM atm) {
        this.bank = bank;
        this.atm = atm;
    }

    public void startSession(String cardNumber, String pin) {
        User user = bank.authenticate(cardNumber, pin);
        if (user == null) {
            System.out.println("Authentication Failed!");
            return;
        }

        Account account = user.getAccount();

        System.out.println("Welcome, " + user.getName() + "!");
        System.out.println("\nAvailable Balance: Rs." + account.getBalance());

        double withdrawAmount = 100.0;
        if (account.withdraw(withdrawAmount) && atm.dispenseCash(withdrawAmount)) {
            System.out.println("\nWithdrawn: Rs." + withdrawAmount);
        } else {
            System.out.println("\nWithdrawal failed!");
        }

        System.out.println("Balance after withdrawal: Rs." + account.getBalance());

        double depositAmount = 50.0;
        account.deposit(depositAmount);
        System.out.println("\nDeposited: Rs." + depositAmount);

        System.out.println("Balance after deposit: Rs." + account.getBalance());
    }
}

File: Low Level Design (LLD)/Problems/01. ATM/Cpp/src/ATMService.cpp
#include <iostream>
#include "../include/ATMService.h"

ATMService::ATMService(Bank* bank, ATM* atm) : bank(bank), atm(atm) {};

void ATMService::startSession(const std::string& cardNumber, const std::string& pin) {
    User* user = bank->authenticate(cardNumber, pin);
    if (!user) {
        std::cout << "Authentication Failed!\n";
        return;
    }

    Account* account = user->getAccount();
    std::cout << "Welcome!" << user->getName() << "\n";
    std::cout << "Initial Balance: $" << account->getBalance() << "\n";

    double withdrawAmount = 100.0;
    if (account->withdraw(withdrawAmount) && atm->dispenseCash(withdrawAmount)) {
        std::cout << "Withdrawn: Rs." << withdrawAmount << "\n";
    } else {
        std::cout << "Withdrawal failed!\n";
    }

    double depositAmount = 50.0;
    account->deposit(depositAmount);
    std::cout << "Deposited: Rs." << depositAmount << "\n";
    std::cout << "Final Balance: Rs." << account->getBalance() << "\n";
}